lib/hutch/broker.rb

```diff
@@ -215,41 +216,8 @@ def nack(delivery_tag)
       channel.nack(delivery_tag, false, false)
     end
 
-    def log_publication(serializer, payload, routing_key)
-      logger.info {
-        spec =
-          if serializer.binary?
-            "#{payload.bytesize} bytes message"
-          else
-            "message '#{payload}'"
-          end
-        "publishing #{spec} to #{routing_key}"
-      }
-    end
-
-    def publish(routing_key, message, properties = {}, options = {})
-      ensure_connection!(routing_key, message)
-
-      serializer = options[:serializer] || @config[:serializer]
-
-      non_overridable_properties = {
-        routing_key:  routing_key,
-        timestamp:    connection.current_timestamp,
-        content_type: serializer.content_type,
-      }
-      properties[:message_id]   ||= generate_id
-
-      payload = serializer.encode(message)
-
-      log_publication(serializer, payload, routing_key)
-
-      response = exchange.publish(payload, {persistent: true}.
-        merge(properties).
-        merge(global_properties).
-        merge(non_overridable_properties))
-
-      channel.wait_for_confirms if @config[:force_publisher_confirms]
-      response
+    def publish(*args)
+      publisher.new(connection, channel, exchange, @config).publish(*args)
```

Comment:
It should only be used on the first publication.

`publisher.publish(*args)` calls the `Broker#publisher` method which just returns `@publisher` if it's set. Otherwise `Broker#publisher` sets and returns `@publisher`.